### ISSUE TITLE
Github integration: API rate limit exceeded

### ISSUE THREAD

**Issue by wookayin** [orhun/git-cliff#422]
First of all, congratulations and thanks for the great work being done on Github integration #363.

**Describe the bug**

On a large repository where there are a lot of commits, API rate limit is always so I can't generate the release note.

My understanding from #363 is that **ALL** of the commits in the repository is fetched, so in this way every git cliff attempt will be turned down by the API limit. Even if # of the commits to be contained in a release note is small (e.g. between two minor releases), it's not like the PR information about each commit is fetched but it's like git-cliff is trying to fetch everything.

**To reproduce**

[neovim/neovim](https://github.com/neovim/neovim) has ~28k commit as of today.

```bash
git clone https://github.com/neovim/neovim && cd neovim

# edit manually `scripts/cliff.toml`, add the following section:
# [remote.github]
# owner = "neovim"
# repo = "neovim"

git cliff --config scripts/cliff.toml v0.9.4..v0.9.5  --github-token <TOKEN>
```

**Expected behavior**

No errors.

**Screenshots / Logs**

Bunch of API limit errors (possibly one log is per page request):

```
▹▹▸▹▹ Retrieving data from GitHub... (neovim/neovim)

 ERROR git_cliff_core::github    > Request error: {
  "documentation_url": "https://docs.github.com/en/free-pro-team@latest/rest/overview/rate-limits-for-the-rest-api#about-secondary-rate-limits",
  "message": "You have exceeded a secondary rate limit. Please wait a few minutes before you try again. If you reach out to GitHub Support for help, please include the request ID D7CB:1218:1333433:289A206:65902213."
}
```

**Software information**

- Operating system: macOS 14
- Rust version: 1.73.0
- Project version: git-cliff 1.4.0 (built from source) after #363 is merged to master

**Comment by orhun** [comment]
Hello! 🐻 

> First of all, congratulations and thanks for the great work being done on Github integration

Thank you! I'm glad it is being tested already before the release.

> Even if # of the commits to be contained in a release note is small (e.g. between two minor releases), it's not like the PR information about each commit is fetched but it's like git-cliff is trying to fetch everything.

Yes, I mentioned the reasoning behind this briefly in #363 and the problem is generating the list of "new contributors". GitHub doesn't provide an API for easily fetching this so we need all the data for figuring out if someone has contributed to the repository in the past.

I don't know how to solve this yet. One thing we can do for large repos is that use date filters for the GitHub API requests so that we only request the relevant part of the history. But then you won't be able to generate the list for first-time contributors - but you can still list the contributors of the release.

Do you think that would be something feasible?

> If I may suggest, fetching PR information for only the relevant commits, e.g., via [api.github.com/repos/neovim/neovim/commits/COMMIT/pulls](https://api.github.com/repos/neovim/neovim/commits/COMMIT/pulls), would be an alternative, efficient way.

This means fetching the pull request for each commit which adds up to a lot of requests as well.

**Comment by wookayin** [comment]
Thanks for the reply.

For me, I don't need "first-time contributors". It would be great if there's a knob for this. If this is disabled, we might be able to save a lot of unnecessary API calls.

**Comment by orhun** [comment]
Yes, I agree. I think I send less API requests based on whether "first-time contributors" variable is used in the template and document this behavior.

**Comment by orhun** [comment]
I took a stab at this at #431, however it doesn't have a significant speed up for the neovim repository. Although `git-cliff` only requests a time frame of the commits/PRs, it still takes a lot of time due to the number of pull requests. Maybe it makes more sense to use the `commit/pulls` API instead as you mentioned. But that approach might also take some time in cases where there are not a lot of PRs but commits.

Any ideas?

> Also, it's not clear how many requests are (unawarely) being made through git cliff. Can we make it a bit more verbose?

You can see the network requests via giving `-v` flag while running `git-cliff.

**Comment by wookayin** [comment]
Thanks, `-v` is helpful. #431 reduces the number of API calls, but if I understood correctly it still tries to fetch **all** the PRs, rather than only in a given range (there are ~30K PRs!) right? I think we might need a fundamentally better approach of not fetching all the (unnecessary) data. Fetching pull information for each of the commits can be also limited, because the number of commits in a big release can be thousands and we probably don't want to make thousands of `pulls` API calls.

**Comment by orhun** [comment]
>  it still tries to fetch all the PRs given a time range right? However, for me it still gives the secondary API limit error on 11-12 pages (which were requested in < 1 seconds) Guess this might be a throttling issue?

Yes, it needs to fetch all the PRs in the given time range (and there is a lot 😕) There is probably a throttling issue.

> Fetching pull information for each of the commits can be also limited, because the number of commits in a big release can be thousands and we probably don't want to make thousands of pulls API calls.

Exactly. :/

**Comment by Roms1383** [comment]
Hi, thanks for the tool, but sadly same error here ^^'

After short investigation I saw you took a stab at implementing it in #431 and wondered why you dropped it.

By the way also noticed [reqwest_middleware::ClientWithMiddleware](https://docs.rs/reqwest-middleware/latest/reqwest_middleware/) present in the codebase, so maybe handle throttling for every request there :)

**Comment by orhun** [comment]
Hey @Roms1383

> After short investigation I saw you took a stab at implementing it in https://github.com/orhun/git-cliff/pull/431 and wondered why you dropped it.

It was probably because the implementation needs to be done for all of the remotes (GitHub, GitLab, etc.) and I didn't have time to prioritize that at the time.

Also another thing to consider is you cannot infer a correct "first contributors" data without fetching everything from the remotes. That API sadly does not exist for most platforms.

> By the way also noticed [reqwest_middleware::ClientWithMiddleware](https://docs.rs/reqwest-middleware/latest/reqwest_middleware/) present in the codebase, so maybe handle throttling for every request there :)

I'm already caching the requests here:

https://github.com/orhun/git-cliff/blob/372af4b90f2e08d7b364f69753fab62cd012d908/git-cliff-core/src/remote/mod.rs#L156-L170

Not sure what you mean by throttling every request 🤔

**Comment by Roms1383** [comment]
Hello @orhun !
Sorry for late reply, I'm swamped ^^

---

> It was probably because the implementation needs to be done for all of the remotes (GitHub, GitLab, etc.) and I didn't have time to prioritize that at the time.

Totally makes sense, I feel you 💯 

---

> Also another thing to consider is you cannot infer a correct "first contributors" data without fetching everything from the remotes. That API sadly does not exist for most platforms.

The repo Im using `git-cliff` for has over a 1,000 commits, maybe that's why, even though I tried to narrow changelog generation down to a specific previous minor (roughly 300 commits), but indeed I didn't know about this limitation 🤔 

---

> I'm already caching the requests here

This already helps a lot 👍 

---

> Not sure what you mean by throttling every request 🤔

What I mean by _request throttling_ is looking at the provider APIs rate limit (e.g.[ Github REST API rate limit](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=latest)) and usually implement a client middleware that keeps track of latest request made to this API, and accordingly throttle or "debounce" future request to send.

So for example if you know that:
- rate limit is [60 requests per hour for unauthenticated user on Gihub](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-unauthenticated-users):
	- you could bail as soon as you know it requires to go well above the quota,
	- maybe incite user to run `git-cliff` as an authenticated user which provides [5000 requests per hour for individuals](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-authenticated-users).

- most of the time each endpoint or kind of endpoints is allowed a secondary requests quota that is e.g.
	- [listing branches on Github](https://docs.github.com/en/rest/branches/branches?apiVersion=2022-11-28#list-branches) consumes [1 point](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#calculating-points-for-the-secondary-rate-limit), and one gets [900 points per minute on Github to spend](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#about-secondary-rate-limits), so you could delay requests over time to avoid crossing the threshold (_throttling_).

- staying below the maximum concurrent requests: e.g. [100 on Github](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#about-secondary-rate-limits).

- usually providers expose some [endpoint(s) to check current rate limit status](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#checking-the-status-of-your-rate-limit), so you can leverage it to make sure to always stay below the quota.

- requests exceeding quota typically return HTTP status code `429` (e.g. [on Github](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit)), so you can also check responses headers to force the client to wait before processing next requests.

- finally some crates can help, e.g. [reqwest_ratelimit](https://docs.rs/reqwest-ratelimit/latest/reqwest_ratelimit/).

Hope this helps a bit :)

**Comment by orhun** [comment]
Thanks a lot for the elaborate reply! This really helped clear up many things :)

I think we can execute a 2 step plan to implement this:

1. Calculate the amount of requests that will be made before sending them and show a warning to user if the ratelimit will be exceeded. This then can be turned into an error after some testing.
2. Implement ratelimit handling for the available remotes and wait a certain amount of time between requests before proceeding. 

How does this sound? Do you have any other suggestions?

**Comment by Roms1383** [comment]
Sounds like a plan ! 💯 
Only suggestion since `git-cliff` is agnostic between code hosting providers is to quickly cross-compare their respective APIs before implementing it to avoid refactoring it later.

**Comment by asweet-confluent** [comment]
> Also another thing to consider is you cannot infer a correct "first contributors" data without fetching everything from the remotes. That API sadly does not exist for most platforms.

At least for GitHub, [the "list commits" API ](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#list-commits--parameters)has a field for filtering by author. If you combine that with `until=<date of first commit in revision range>`, you should only have to make one request per author in the changelog.

**Comment by Jefffrey** [comment]
I'm interested to see if we can optimize it to make this tool go even faster 👀

What I understand from the current codebase:

- Fetches all PRs because for the range of release commits, needs to search for associated PR (if exists)
- Fetches all commits, because commits before the range of release are used for first time contributor statistics

## Commits

As pointed out above, there is an API endpoint for listing commits by author, though this means firing a query per author. What I have in mind is a hybrid approach:

- First fetch all commits from GitHub in the release range (required)
- Get unique list of contributors
- If unique list of contributors is less than no. of API requests needed to get other historical commits (aka no. of commits to fetch / page size (100)) then use per author query fetching
- Otherwise fall back to just fetching all commits (as before)

I think this would give a best case of reducing the number of requests since usually the no. of unique contributors is quite low. And worst case (unique contributor every commit lmao) we won't get an explosion of requests because of the fallback.

- Edit: we could do a hybrid approach like PRs below, fetching batches of commits at a time until the no. of unique contributors is small enough to fire individual requests for their commits)

## PRs

We first gather a unique set of the commit sha's we are looking for; then we call the list PR API ensuring we specify a sort on updated (I assume this is better than created?), with direction of descending. As we process each page, we try to find the match in the set and remove as we go. We keep this going until we get to a low enough amount in the set that we can just start using the API that finds the PR for a commit, [List pull requests associated with a commit](https://docs.github.com/en/rest/commits/commits?apiVersion=2026-03-10&versionId=free-pro-team%40latest&productId=rest#list-pull-requests-associated-with-a-commit). Similar trade off as above, where we determine this threshold based on if its less requests to just get the rest of the PRs, or to do this per commit approach.

This way even if in worst case there are commits with no associated PRs, we can still account for looking them up without needing to get the entire history.

## Feasible?

This is just theoretical, maybe there are some nuances to the API that might break some of my assumptions above. Also it would break the current flow thats standardized across remotes of doing `fetch -> process`, since it would intertwine these together. Just curious if such an approach sounds good to try and implement?

Edit: this would likely be applicable only when there's one release being processed, for simplicity

**Comment by orhun** [comment]
Hey @Jefffrey, that sounds good to me, feel free to drop a PR for it :)
